src/main/scala/mesosphere/marathon/upgrade/DeploymentPlan.scala

```diff
@@ -59,13 +59,164 @@ final case class DeploymentPlan(
     steps: Seq[DeploymentStep],
     version: Timestamp) extends MarathonState[Protos.DeploymentPlanDefinition, DeploymentPlan] {
 
+  import DeploymentPlan.log
+
   def isEmpty: Boolean = steps.isEmpty
 
   def nonEmpty: Boolean = !isEmpty
 
+  /** @return all ids of apps which are referenced in any deployment actions */
   def affectedApplicationIds: Set[PathId] = steps.flatMap(_.actions.map(_.app.id)).toSet
 
+  /**
+    * Reverts this plan by applying the reverse changes to the given Group.
+    *
+    * If they were no concurrent deployments and the passed group is equal to
+    * the target group, this method should result in the original group
+    * except for version/timestamp changes.
+    *
+    * If there were concurrent changes, this method tries to revert the changes
+    * of this deployment only without affecting the other deployments.
+    */
+  val revert: Group => Group = (revertGroupChanges _).andThen(revertAppChanges)
+
+  /**
+    * Reverts any group additions, changes, removals.
+    *
+    * It is more difficult than reverting any app definition changes
+    * because groups are not locked by deployments and concurrent changes are allowed.
+    */
+  private[this] def revertGroupChanges(group: Group): Group = {
+    // sort groups so that inner groups are processed first
+    val sortedGroupChanges = groupChanges.sortWith {
+      case (change1, change2) =>
+        // both groups are supposed to have the same path id (if there are any)
+        def pathId(change: (Option[Group], Option[Group])): PathId = {
+          Seq(change._1, change._2).flatten.map(_.id).headOption.getOrElse(PathId.empty)
+        }
+
+        pathId(change1) > pathId(change2)
+    }
+
+    sortedGroupChanges.foldLeft(group) {
+      case (result, groupUpdate) =>
+        groupUpdate match {
+          case (Some(oldGroup), None) => // revert group removal
```

Comment:
There's lots of code here. Would it make sense to create methods for the individual cases to increase readability?